product_pricelist_supplierinfo/models/product_product.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2018 Tecnativa - Vicent Cubells
+# Copyright 2018 Tecnativa - Pedro M. Baeza
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, models, tools
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    def _get_supplierinfo_pricelist_price(self, rule, date=None,
+                                          quantity=None):
+        """Method for getting the price from supplier info."""
+        self.ensure_one()
+        domain = [
+            '|',
+            ('product_id', '=', self.id),
+            ('product_tmpl_id', '=', self.product_tmpl_id.id),
+        ]
+        if not rule.no_supplierinfo_min_quantity and quantity:
+            domain += [
+                '|',
+                ('min_qty', '=', False),
+                ('min_qty', '>=', quantity),
+            ]
+        if date:
+            domain += [
+                '|',
+                ('date_start', '=', False),
+                ('date_start', '<=', date),
+                '|',
+                ('date_end', '=', False),
+                ('date_end', '>=', date),
+            ]
+        # We use a different default order because we are interested in getting
+        # the price for lowest minimum quantity
+        price = self.env['product.supplierinfo'].search(
```

Comment:
@pedrobaeza what prevents us from doing this search for all products at once and then doing the loop? (nitpicking :smirk)
In a lot of cases, price is computed for a lot of product with qty=1. The Idea would be to write the domain with 
```python
            '|',
            ('product_id', 'in', self.ids),
            ('product_tmpl_id', 'in', self.product_tmpl_id.ids),
```
 order the search by `order= 'product_id, product_tmpl_id, sequence, min_qty, price'`

The price can be computed each time the product change into the loop.

```python
result = dict.from_keys(self.ids, None)
psis = self.env['product.supplierinfo'].search(domain, order= 'product_id, product_tmpl_id, sequence, min_qty, price')
if psis:
    product_id = psis[0].product_id
    for psi in psis:
        if product_id != psi.product_id:
            # here we put all the logic from line 41 to line 62
            ...
            result[product_id.id] = price
return result
```

Reply:
I don't think this deserves to be done as too convoluted: you come from the other method where each argument is a tuple that have to be evaluated for the proper rule. Then we would need to store all products and their corresponding quantities, and call this multi-record aware method with that new structure, and deal with this properly.

Uf, too much effort being this call normally call for one record when retrieving the price in the sales order line.

Reply:
Exactly, that's the main point I think: optimizing this will have 0 benefit with the super code we are already handling. Take into account that I'm already dealing with several passes over same tuple list for recomputing what is not correct, but before there has been 2 useless computations already. We have to push for a proper refactoring on master.